hal/halx86/acpi/madt.c

```diff
@@ -12,77 +13,98 @@
     _In_ PLOADER_PARAMETER_BLOCK LoaderBlock)
 {
+    ACPI_TABLE_MADT *MadtTable;
+    ACPI_SUBTABLE_HEADER *AcpiHeader;
     ULONG_PTR TableEnd;
-    ULONG ValidProcessorCount;
-
-    /* We only support legacy APIC for now, this will be updated in the future */
-    HalpApicInfoTable.ApicMode = 0x10;
-    MadtTable = HalAcpiGetTable(LoaderBlock, 'CIPA');
 
-    AcpiHeader = (ACPI_SUBTABLE_HEADER*)MadtTable;
-    AcpiHeader->Length = sizeof(ACPI_TABLE_MADT);
-    TableEnd = (ULONG_PTR)MadtTable + MadtTable->Header.Length;
+    MadtTable = HalAcpiGetTable(LoaderBlock, APIC_SIGNATURE);
+    if (!MadtTable)
+    {
+        DPRINT1("HalAcpiGetTable(APIC) failed\n");
+        return;
+    }
 
-    HalpApicInfoTable.ProcessorCount = 0;
-    HalpProcessorIdentity = HalpStaticProcessorIdentity;
+    /* TODO: We only support legacy APIC for now */
+    HalpApicInfoTable.ApicMode = APIC_MODE_LEGACY;
 
-    AcpiHeader = (ACPI_SUBTABLE_HEADER*)MadtTable;
-    AcpiHeader->Length = sizeof(ACPI_TABLE_MADT);
+    AcpiHeader = (ACPI_SUBTABLE_HEADER *)((ULONG_PTR)MadtTable + sizeof(*MadtTable));
     TableEnd = (ULONG_PTR)MadtTable + MadtTable->Header.Length;
+    DPRINT("MadtTable %p, subtables %p - %08Ix\n", MadtTable, AcpiHeader, TableEnd);
 
-    while ((ULONG_PTR)AcpiHeader <= TableEnd)
+    while ((ULONG_PTR)AcpiHeader < TableEnd)
     {
-        LocalApic = (ACPI_MADT_LOCAL_APIC*)AcpiHeader;
-
-        if (LocalApic->Header.Type == ACPI_MADT_TYPE_LOCAL_APIC &&
-            LocalApic->Header.Length == sizeof(ACPI_MADT_LOCAL_APIC))
+        switch (AcpiHeader->Type)
         {
-            ValidProcessorCount = HalpApicInfoTable.ProcessorCount;
-
-            HalpProcessorIdentity[ValidProcessorCount].LapicId = LocalApic->Id;
-            HalpProcessorIdentity[ValidProcessorCount].ProcessorId = LocalApic->ProcessorId;
-
-            HalpApicInfoTable.ProcessorCount++;
-
-            AcpiHeader = (ACPI_SUBTABLE_HEADER*)((ULONG_PTR)AcpiHeader + AcpiHeader->Length);
-        }
-        else
-        {
-            /* End the parsing early if we don't use the currently selected table */
-            AcpiHeader = (ACPI_SUBTABLE_HEADER*)((ULONG_PTR)AcpiHeader + 1);
+            case ACPI_MADT_TYPE_LOCAL_APIC:
+            {
+                ACPI_MADT_LOCAL_APIC *LocalApic = (ACPI_MADT_LOCAL_APIC *)AcpiHeader;
+
+                DPRINT(" LocalApic, Processor %lu: ProcessorId %u, Id %u, LapicFlags %08X\n",
+                       HalpApicInfoTable.ProcessorCount,
+                       LocalApic->ProcessorId, LocalApic->Id, LocalApic->LapicFlags);
+
+                HalpProcessorIdentity[HalpApicInfoTable.ProcessorCount].ProcessorId = LocalApic->ProcessorId;
+                HalpProcessorIdentity[HalpApicInfoTable.ProcessorCount].LapicId = LocalApic->Id;
+#if 0 // TODO: Is that correct?
+                if (LocalApic->LapicFlags & 0x00000001)
+                    HalpProcessorIdentity[HalpApicInfoTable.ProcessorCount].ProcessorStarted = TRUE;
+#endif
```

Comment:
Please, check this addition, as I'm not sure when/how `ProcessorStarted` is supposed to be set.

Reply:
This is suppose to be set the processor when it's in kernel mode and we're Initializing HAL.

Of course we haven't quite gotten there yet in master. 

But that's what it's for.

Reply:
Thanks, I'll replace this wrong code.

While here, when/how is `BSPCheck` supposed to be set?

Reply:
It's not merged yet but it's setup the processor that started the system

Reply:
Thanks, I'll leave that alone.